src/librustc_trait_selection/traits/project.rs

```diff
@@ -1258,6 +1265,42 @@ fn confirm_generator_candidate<'cx, 'tcx>(
         .with_addl_obligations(obligations)
 }
 
+fn confirm_discriminant_kind_candidate<'cx, 'tcx>(
+    selcx: &mut SelectionContext<'cx, 'tcx>,
+    obligation: &ProjectionTyObligation<'tcx>,
+    discriminant_kind_vtable: VtableDiscriminantKindData<'tcx, PredicateObligation<'tcx>>,
+) -> Progress<'tcx> {
+    let tcx = selcx.tcx();
+
+    let enum_ty = selcx.infcx().shallow_resolve(discriminant_kind_vtable.enum_ty);
+    let substs = tcx.mk_substs([enum_ty.into()].iter());
+
+    let assoc_items = tcx.associated_items(tcx.lang_items().discriminant_kind_trait().unwrap());
+    // FIXME: emit an error if the trait definition is wrong
+    let discriminant_def_id = assoc_items.in_definition_order().next().unwrap().def_id;
+
+    let discriminant_ty = match enum_ty.kind {
+        // Use the discriminant type for enums
+        ty::Adt(adt, _) if adt.is_enum() => adt.repr.discr_type().to_ty(tcx),
+        // Default to `i32` for generators.
+        //
+        // FIXME: We may want to use the smallest
+        // fitting integer type instead.
```

Comment:
We don't, this is the high-level discriminant, not the low-level *tag*. We shouldn't even be using a signed integer, but I guess that's a preexisting bug with generators.

Reply:
cc @tmandry

Reply:
I disagree. The high-level discriminant is used in most derives for enums. Using `u8` instead of `u64` is most probably a noticeable performance improvement

Reply:
This is for generators, they don't have derives. Also, integer size doesn't matter for *by-value* things, only for memory usage or vectorization (i.e. when you have arrays of integers).

Reply:
> This is for generators, they don't have derives.

:thinking: :thinking: :thinking: :thinking: :thinking: